File: 1456-maximum-number-of-vowels-in-a-substring-of-given-length/1456-maximum-number-of-vowels-in-a-substring-of-given-length.py
class Solution:
    def maxVowels(self, s, k):
        
        vowels = set('aeiou')
        
        # Step 1: first window
        count = 0
        for i in range(k):
            if s[i] in vowels:
                count += 1
        
        max_count = count
        
        # Step 2: slide window
        for i in range(k, len(s)):
            
            # add new character
            if s[i] in vowels:
                count += 1
            
            # remove old character
            if s[i-k] in vowels:
                count -= 1
            
            max_count = max(max_count, count)
        
        return max_count

File: 1456-maximum-number-of-vowels-in-a-substring-of-given-length/1456-maximum-number-of-vowels-in-a-substring-of-given-length.java
class Solution {
    public int maxVowels(String s, int k) {
        int maxCount = 0;
        int count = 0;
        
        // Step 1: initial window
        for (int i = 0; i < k; i++) {
            if (isVowel(s.charAt(i))) {
                count++;
            }
        }
        
        maxCount = count;
        
        // Step 2: sliding window
        for (int i = k; i < s.length(); i++) {
            // add new character
            if (isVowel(s.charAt(i))) {
                count++;
            }
            
            // remove old character
            if (isVowel(s.charAt(i - k))) {
                count--;
            }
            
            maxCount = Math.max(maxCount, count);
        }
        
        return maxCount;
    }
    
    private boolean isVowel(char c) {
        return c == 'a' || c == 'e' || c == 'i' || 
               c == 'o' || c == 'u';
    }
}